Upgrading the versions of images in Dockerfile.

wanghuaiqing2010 opened the pull request:
In order to run tests at mips64el device.
Now official-images has supported the following images for mips64el.
buildpack-deps:stretch
buildpack-deps:buster
debian:stretch
debian:buster

But official-images does not support the following images for mips64el.
debian:jessie
buildpack-deps:jessie

Signed-off-by: wanghuaiqing <wanghuaiqing@loongson.cn>

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah commented:
Seeing some failures in CI;

On amd64, ppc64le, s390x, and amd64 (rootless):

```
`=== RUN   TestNetworkLoopbackNat
--- FAIL: TestNetworkLoopbackNat (11.03s)
    nat_test.go:76: timeout hit after 10s: waiting for container to be stopped
```

On amd64:

```
`=== RUN   TestDockerSuite/TestContainerAPITop
    --- FAIL: TestDockerSuite/TestContainerAPITop (0.60s)
        docker_api_containers_test.go:413: assertion failed: 1 (int) != 2 (int): expected 2 processes, found 1: [[root 26387 0.0 0.0 1300 4 ? Ss 07:18 0:00 top]]


=== RUN   TestDockerDaemonSuite/TestDaemonUlimitDefaults
    --- FAIL: TestDockerDaemonSuite/TestDaemonUlimitDefaults (1.42s)
        docker_cli_daemon_test.go:976: expected `ulimit -p` to be 2048, got:
```


I'll restart CI to see if the last two are possibly flaky

thaJeztah commented:
I opened https://github.com/moby/moby/pull/41147 to remove the outdated example dockerfiles

wanghuaiqing2010 commented on contrib/desktop-integration/chromium/Dockerfile:
```diff
@@ -21,7 +21,7 @@
 
 # Base docker image
-FROM debian:jessie
+FROM debian:buster
```
Yes, I have merged your changes.

wanghuaiqing2010 commented on contrib/desktop-integration/gparted/Dockerfile:
```diff
@@ -18,7 +18,7 @@
 
 # Base docker image
-FROM debian:jessie
+FROM debian:buster
```
Yes, I have merged your changes.

wanghuaiqing2010 commented:
There is a difference in the behavior of the program “sh" between the old busybox and new busybox.
This will cause the following problem, so I have modified the expected results.

=== RUN   TestDockerSuite/TestContainerAPITop
    --- FAIL: TestDockerSuite/TestContainerAPITop (0.60s)
        docker_api_containers_test.go:413: assertion failed: 1 (int) != 2 (int): expected 2 processes, found 1: [[root 26387 0.0 0.0 1300 4 ? Ss 07:18 0:00 top]]

wanghuaiqing2010 commented:
There is a difference in the option of the program “ulimit" between the old busybox and new busybox.
This will cause the following problem.

=== RUN   TestDockerDaemonSuite/TestDaemonUlimitDefaults
    --- FAIL: TestDockerDaemonSuite/TestDaemonUlimitDefaults (1.42s)
        docker_cli_daemon_test.go:976: expected `ulimit -p` to be 2048, got:

I have made the following modifications to solve this problem.
```
diff --git a/integration-cli/docker_cli_daemon_test.go b/integration-cli/docker_cli_daemon_test.go
index 41b0212676..44d7026bb9 100644
--- a/integration-cli/docker_cli_daemon_test.go
+++ b/integration-cli/docker_cli_daemon_test.go
@@ -957,7 +957,7 @@ func (s *DockerDaemonSuite) TestDaemonUlimitDefaults(c *testing.T) {

        s.d.StartWithBusybox(c, "--default-ulimit", "nofile=42:42", "--default-ulimit", "nproc=1024:1024")

-       out, err := s.d.Cmd("run", "--ulimit", "nproc=2048", "--name=test", "busybox", "/bin/sh", "-c", "echo $(ulimit -n); echo $(ulimit -p)")
+       out, err := s.d.Cmd("run", "--ulimit", "nproc=2048", "--name=test", "busybox", "/bin/sh", "-c", "echo $(ulimit -n); echo $(ulimit -u)")
        if err != nil {
                c.Fatal(err, out)
        }
@@ -973,7 +973,7 @@ func (s *DockerDaemonSuite) TestDaemonUlimitDefaults(c *testing.T) {
                c.Fatalf("expected `ulimit -n` to be `42`, got: %s", nofile)
        }
        if nproc != "2048" {
-               c.Fatalf("expected `ulimit -p` to be 2048, got: %s", nproc)
+               c.Fatalf("expected `ulimit -u` to be 2048, got: %s", nproc)
        }
```

thaJeztah commented:
hmm. interesting @tianon that ring a bell?

wanghuaiqing2010 commented:
There is a difference on semantics of the option "-w" of the program "nc" between the old busybox and new busybox.

1）The new busybox:
-w SEC	Timeout for connects and final net reads

2）The old busybox:
-w SEC	Timeout for connects

This will cause the following problem, so I have modified timeout to 1 second.
`=== RUN   TestNetworkLoopbackNat
--- FAIL: TestNetworkLoopbackNat (11.03s)
    nat_test.go:76: timeout hit after 10s: waiting for container to be stopped

wanghuaiqing2010 commented:
The new busybox “nslookup” takes longer to resolve domain names.
It will cause the following problem.
```
=== RUN   TestDaemonDNSFallback
--- FAIL: TestDaemonDNSFallback (8.30s)
    dns_test.go:34: timeout hit after 4s: waiting for container to be "exited", currently running
```
So I have made the following changes to solve the problem.
```

diff --git a/integration/network/dns_test.go b/integration/network/dns_test.go
index 19d954afea..f8037f0184 100644
--- a/integration/network/dns_test.go
+++ b/integration/network/dns_test.go
@@ -31,5 +31,5 @@ func TestDaemonDNSFallback(t *testing.T) {
        cid := container.Run(ctx, t, c, container.WithNetworkMode("test"), container.WithCmd("nslookup", "docker.com"))
        defer c.ContainerRemove(ctx, cid, types.ContainerRemoveOptions{Force: true})

-       poll.WaitOn(t, container.IsSuccessful(ctx, c, cid), poll.WithDelay(100*time.Millisecond), poll.WithTimeout(4*time.Second))
+       poll.WaitOn(t, container.IsSuccessful(ctx, c, cid), poll.WithDelay(100*time.Millisecond), poll.WithTimeout(10*time.Second))
 }
```

tianon commented:
I believe the `ulimit` change occurred in BusyBox 1.31.0:

> `ash,hush: bash compat for ulimit: -w => -x, -p => -u`

tianon commented:
Usage of the `nslookup` utility is discouraged by BusyBox upstream maintainers -- not sure why it would be taking longer though the underlying implementation did change (in many ways not for the better -- the k8s community was particularly up-in-arms about it). :disappointed:

tianon commented on integration-cli/docker_api_containers_test.go:
```diff
@@ -410,9 +410,8 @@ func (s *DockerSuite) TestContainerAPITop(c *testing.T) {
 		c.Fatalf("expected `USER` at `Titles[0]` and `COMMAND` at Titles[10]: %v", top.Titles)
 	}
-	assert.Equal(c, len(top.Processes), 2, fmt.Sprintf("expected 2 processes, found %d: %v", len(top.Processes), top.Processes))
-	assert.Equal(c, top.Processes[0][10], "/bin/sh -c top")
```
This looks like an optimization in `/bin/sh` -- it could probably be "fixed" by using `"/bin/sh -c 'top && true'"` or similar (to force `/bin/sh` to stick around), if it's really important for this test to have multiple processes.

wanghuaiqing2010 replied:
Yes，I have modified the code as you requested. It works.
```
diff --git a/integration-cli/docker_api_containers_test.go b/integration-cli/docker_api_containers_test.go
index 3a30ea8..f6ea1fc 100644
--- a/integration-cli/docker_api_containers_test.go
+++ b/integration-cli/docker_api_containers_test.go
@@ -394,7 +394,7 @@ func (s *DockerSuite) TestContainerAPIPause(c *testing.T) {
 
 func (s *DockerSuite) TestContainerAPITop(c *testing.T) {
        testRequires(c, DaemonIsLinux)
-       out, _ := dockerCmd(c, "run", "-d", "busybox", "/bin/sh", "-c", "top")
+       out, _ := dockerCmd(c, "run", "-d", "busybox", "/bin/sh", "-c", "top && true")
        id := strings.TrimSpace(out)
        assert.NilError(c, waitRun(id))
 
@@ -411,7 +411,7 @@ func (s *DockerSuite) TestContainerAPITop(c *testing.T) {
                c.Fatalf("expected `USER` at `Titles[0]` and `COMMAND` at Titles[10]: %v", top.Titles)
        }
        assert.Equal(c, len(top.Processes), 2, fmt.Sprintf("expected 2 processes, found %d: %v", len(top.Processes), top.Processes))
-       assert.Equal(c, top.Processes[0][10], "/bin/sh -c top")
+       assert.Equal(c, top.Processes[0][10], "/bin/sh -c top && true")
        assert.Equal(c, top.Processes[1][10], "top")
 }
```

Commit:
Upgrading the versions of images in Dockerfile.

In order to run tests at mips64el device.
Now official-images has supported the following images for mips64el.
buildpack-deps:stretch
buildpack-deps:buster
debian:stretch
debian:buster

But official-images does not support the following images for mips64el.
debian:jessie
buildpack-deps:jessie

Signed-off-by: wanghuaiqing <wanghuaiqing@loongson.cn>

thaJeztah reviewed (approved):
LGTM

thaJeztah commented:
@tianon ptal

tianon reviewed (approved):
LGTM :+1:

thaJeztah commented:
Thanks, @wanghuaiqing2010 !
